# Sparse Learning, Distance and (k)NN method

> Weitong Zhang
> 2015011493
>
> <zwt15@mails.tsinghua.edu.cn>

## Voronor Gird in Euclid Space

We are about to prove the following statement:

$$\begin{cases}
\forall i\ \|\vec x_i - \vec s_1 \|_2^2 \ge \|\vec x_0 - \vec s_1 \|_2^2\\
\forall i\ \|\vec x_i - \vec s_2 \|_2^2 \ge \|\vec x_0 - \vec s_2 \|_2^2\\
\forall t\ \in [0,1], \vec s = t\vec s_1 + (1-t) \vec s_2
\end{cases} \Rightarrow \forall i\ \|\vec x_i - \vec s \|_2^2 \ge \|\vec x_0 - \vec s \|_2^2
$$

$$\begin{cases}
\|\vec x_i - \vec s \|_2^2 - \|\vec x_0 - \vec s \|_2^2 = \sum_j (x_{ij} - s_j)^2 - \sum_j (x_{0j} - s_j)^2 = \sum_j (x_{ij} - x_{0j})(x_{ij} + x_{0j} - 2s_j)\\
\|\vec x_i - \vec s_1 \|_2^2 - \|\vec x_0 - \vec s_1 \|_2^2 = \sum_j (x_{ij} - s_{1j})^2 - \sum_j (x_{0j} - s_{1j})^2 = \sum_j (x_{ij} - x_{0j})(x_{ij} + x_{0j} - 2s_{1j}) \ge 0\\
\|\vec x_i - \vec s_2 \|_2^2 - \|\vec x_0 - \vec s_2 \|_2^2 = \sum_j (x_{ij} - s_{2j})^2 - \sum_j (x_{0j} - s_{2j})^2 = \sum_j (x_{ij} - x_{0j})(x_{ij} + x_{0j} - 2s_{2j}) \ge 0
\end{cases}$$

We can easily found out that the first inequation is the linear combination of the later two, therefore, 

$$ \begin{aligned}
\|\vec x_i - \vec s \|_2^2 - \|\vec x_0 - \vec s \|_2^2 = \sum_j (x_{ij} - s_j)^2 - \sum_j (x_{0j} - s_j)^2 = \sum_j (x_{ij} - x_{0j})(x_{ij} + x_{0j} - 2s_j) \\
= t(\sum_j (x_{ij} - x_{0j})(x_{ij} + x_{0j} - 2s_{1j})) + (1-t) (\sum_j (x_{ij} - x_{0j})(x_{ij} + x_{0j} - 2s_{2j})) \ge 0
\end{aligned}$$

Therefore, the set determined by Voronoi is a convex set

## Error rate of NN method

### Error rate of Baysian Method

Since the probability of $x \in [0,\frac{cr}{c-1}]$, the classifier could be only randomly chosen from all of the $c$ categories

$$P^* = \sum_i P(w_i)P_{err} = \frac{cr}{c-1} \times \frac{c-1}{c} = r$$

### Error rate of NN method

Suppose that the number of samples in the training dataset is sufficient, i.e. for each $x$ where $p(x|w_i) \ne 0$, there are enough samples belong to $w_i$

For each $x\in w_i$, if $x \in [i,i+ 1 - \frac{cr}{c-1}]$, the NN method will not generate error, if $x\in [0,\frac{cr}{c-1}]$, the nearest sample of $x$ belongs to all of the $c$ categories, therefore, the probability of error is $\frac{c-1}{c}$

Therefore, the error rate of NN method is 

$$\frac{cr}{c-1} \times \frac{c-1}{c} = r = P^*$$

## Minkowski Distance

Minkowski distance should be described as:

$$ D(X,Y) = (\sum_i |x_i - y_i|^p )^{1/p}$$

According to the definition of distance, a distance should obey the following rules:

$$\begin{aligned}
D(p,q) & \ge 0, D(p,q) = 0 \Leftrightarrow p = q\\
D(p,q) &= D(q,p)\\
D(p,q) &\le D(p,z) + D(q,z)
\end{aligned}$$

Owing to the fact that $|x| = |-x| \ge 0, |x| = 0 \Leftrightarrow x = 0$, the first two rules are easily satisfied. We are about to prove:

$$ (\sum_i |x_i + y_i|^p )^{1/p} \le (\sum_i |x_i|^p )^{1/p} + (\sum_i |y_i|^p )^{1/p}$$

$$ \sum_i |x_i + y_i|^p = \sum_i |x_i + y_i| \times |x_i + y_i|^{p-1} \le \sum_i |x_i| \times |x_i + y_i|^{p-1} + \sum_i |y_i| \times |x_i + y_i|^{p-1}$$

Now, we have to use the $\mathrm {H\ddot older}$ inequation:

$$\sum_i u_iv_i \le (\sum u_i^p)^{\frac1p}(v_i^q)^{\frac1q}, \text{ where } \frac1p + \frac1q = 1$$

Therefore:

$$\sum_i |x_i| \times |x_i + y_i|^{p-1} \le (\sum_i |x_i|^p)^{\frac1p}(\sum_i|x_i + y_i|^{(p-1)q})^{\frac1q}, \text{ where } \frac1p + \frac1q = 1 \Rightarrow pq - q = p$$

Therefore:

$$\sum_i |x_i| \times |x_i + y_i|^{p-1} + \sum_i |y_i| \times |x_i + y_i|^{p-1} \le (\sum_i |x_i|^p)^{\frac1p}(\sum_i |x_i + y_i|^p)^{\frac1q} + (\sum_i |y_i|^p)^{\frac1p}(\sum_i |x_i + y_i|^p)^{\frac1q}$$

Therefore, since $\frac1p + \frac1q = 1$, we get

$$\begin{aligned}
&\sum_i |x_i + y_i|^p \le ((\sum_i |y_i|^p)^{\frac1p} + (\sum_i |x_i|^p)^{\frac1p})(\sum_i |x_i + y_i|^p)^{\frac1q}\\
&\Leftrightarrow \frac{\sum_i |x_i + y_i|^p}{(\sum_i |x_i + y_i|^p)^{\frac1q}} \le (\sum_i |y_i|^p)^{\frac1p} + (\sum_i |x_i|^p)^{\frac1p}\\
&\Leftrightarrow (\sum_i |x_i + y_i|^p )^{1/p} \le (\sum_i |x_i|^p )^{1/p} + (\sum_i |y_i|^p )^{1/p}
\end{aligned}$$

The prove above use the [Hölder's inequality](https://en.wikipedia.org/wiki/Hölder%27s_inequality#Proof_of_Hölder's_inequality)


## Programming

### Getting the MNIST data

In order to minimize the file to upload, we get the MNIST data from website each time

In [1]:
import gzip
import matplotlib.pyplot as plt
import numpy as np
import os
import shutil
import struct
import sys
try: 
    from urllib.request import urlretrieve 
except ImportError: 
    from urllib import urlretrieve
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from scipy.stats import mode
import time

In [2]:
def loadData(src, cimg):
    print ('Downloading ' + src)
    gzfname, h = urlretrieve(src, './delete.me')
    try:
        with gzip.open(gzfname) as gz:
            n = struct.unpack('I', gz.read(4))
            # Read magic number.
            if n[0] != 0x3080000:
                raise Exception('Invalid file: unexpected magic number.')
            # Read number of entries.
            n = struct.unpack('>I', gz.read(4))[0]
            if n != cimg:
                raise Exception('Invalid file: expected {0} entries.'.format(cimg))
            crow = struct.unpack('>I', gz.read(4))[0]
            ccol = struct.unpack('>I', gz.read(4))[0]
            if crow != 28 or ccol != 28:
                raise Exception('Invalid file: expected 28 rows/cols per image.')
            # Read data.
            res = np.frombuffer(gz.read(cimg * crow * ccol), dtype = np.uint8)
    finally:
        os.remove(gzfname)
    return res.reshape((cimg, crow * ccol))

def loadLabels(src, cimg):
    print ('Downloading ' + src)
    gzfname, h = urlretrieve(src, './delete.me')
    try:
        with gzip.open(gzfname) as gz:
            n = struct.unpack('I', gz.read(4))
            # Read magic number.
            if n[0] != 0x1080000:
                raise Exception('Invalid file: unexpected magic number.')
            # Read number of entries.
            n = struct.unpack('>I', gz.read(4))
            if n[0] != cimg:
                raise Exception('Invalid file: expected {0} rows.'.format(cimg))
            # Read labels.
            res = np.frombuffer(gz.read(cimg), dtype = np.uint8)
    finally:
        os.remove(gzfname)
    return res.reshape((cimg, 1))

def try_download(dataSrc, labelsSrc, cimg):
    data = loadData(dataSrc, cimg)
    labels = loadLabels(labelsSrc, cimg)
    return data.astype(np.float32)/256.0,labels

# URLs for the train image and label data
url_train_image = 'http://yann.lecun.com/exdb/mnist/train-images-idx3-ubyte.gz'
url_train_labels = 'http://yann.lecun.com/exdb/mnist/train-labels-idx1-ubyte.gz'
num_train_samples = 60000
train_data,train_labels = try_download(url_train_image, url_train_labels, num_train_samples)

# URLs for the test image and label data
url_test_image = 'http://yann.lecun.com/exdb/mnist/t10k-images-idx3-ubyte.gz'
url_test_labels = 'http://yann.lecun.com/exdb/mnist/t10k-labels-idx1-ubyte.gz'
num_test_samples = 10000
test_data,test_labels = try_download(url_test_image, url_test_labels, num_test_samples)

### Kernel of NN method

In [3]:
train_sum = 60000
test_sum = 10000

def kernel_nn(k,distance,testing,data,label):
    dist = np.hstack((distance(testing,data).reshape((-1,1)),label))
    sel = dist[np.lexsort(dist[:,::-1].T)]
    sel = sel[:k,1]
    predict,_ = mode(sel)
    return predict.astype(np.uint8)

def minkowski(test,data, p = 2):
    r_pos = data-test.reshape((1,-1))
    return np.linalg.norm(r_pos,axis = 1,ord = p)
    
def main_nn(size,k,distance):
    data_used, _, label_used, _ = train_test_split(train_data,train_labels,test_size = train_sum - size)
    predict = []
    start = time.clock()
    for testing in test_data:
        predict.append(kernel_nn(k,distance, testing,data_used,label_used))
    end = time.clock()
    return accuracy_score(test_labels,predict), end - start

### The influence of number of samples on the performance of the classifier

According to the algorithm, the space complexity is $\mathcal O(n)$ (suppose the sort function could be done locally). The time complexity is $\mathcal O(N\log N) + \mathcal O(Nd)$, where $N$ is the number of samples. $d$ is the dimension of the features.

The accuracy of the classifier would keep increasing as the number of the samples increasing. However, the increasing rate would be slow down

In this experiment, we set $k = 5$ and the distance function is euclid distance.

In [4]:
for samples in [6,60,600,6000,60000]:
    preform,t = main_nn(samples,5,minkowski)
    print ('Sample = {}, Acc = {:.1f}%, TimeElapsed = {:e}ms'.format(samples,preform * 100, t * 1000))

Sample = 6, Acc = 10.4%, TimeElapsed = 3.056995e+03ms
Sample = 60, Acc = 56.5%, TimeElapsed = 3.510983e+03ms
Sample = 600, Acc = 84.4%, TimeElapsed = 1.202967e+04ms
Sample = 6000, Acc = 93.7%, TimeElapsed = 5.651620e+05ms
Sample = 60000, Acc = 96.9%, TimeElapsed = 9.257722e+06ms


### The influence of $k$ on the performance of the classifier

We test $k = 1,3,5,7,9,11$ in the situation that number of samples is 60,000 to check the incluence of $k$ on the performance of the classifier. The distance function is still set to euclid function.

As the result shows, the accuracy of the performance is not increasing, This might because the number of samples is too large, therefore, we do this experiment again setting the number of samples to 600.

After setting the number of samples to 600, we can find out that the accuracy of the classifier is dropping, which might because the train set is too small, that the 11-th sample, 9-th sample are not belong to the same category of the testing one.

In [ ]:
for k in [1,3,5,7,9,11]:
    preform,t = main_nn(60000,k,minkowski)
    print ('k = {}, Acc = {:.1f}%, TimeElapsed = {:e}ms'.format(k,preform * 100, t * 1000))

In [6]:
for k in [1,3,5,7,9,11]:
    preform,t = main_nn(600,k,minkowski)
    print ('k = {}, Acc = {:.1f}%, TimeElapsed = {:e}ms'.format(k,preform * 100, t * 1000))

k = 1, Acc = 86.7%, TimeElapsed = 1.157651e+04ms
k = 3, Acc = 84.3%, TimeElapsed = 1.159494e+04ms
k = 5, Acc = 85.2%, TimeElapsed = 1.151041e+04ms
k = 7, Acc = 84.5%, TimeElapsed = 1.178855e+04ms
k = 9, Acc = 82.5%, TimeElapsed = 1.183797e+04ms
k = 11, Acc = 82.8%, TimeElapsed = 1.196463e+04ms


### Using different distance function

We use minkowski distance, where $p = 1$ (Manhattan Distance), $p = 2$ (Euclid Distance, mentioned above), Chebyshev distance to carry on this experiment.

In order to improve the calc speed, we use 6000 samples only (The accuracy of euclid distance could raise up to 93.7%)

In [15]:
def manhattan(test,data):
    return minkowski(test,data,p=1)

def chebyshev(test,data):
    r_pos = data-test.reshape((1,-1))
    return np.max(r_pos,axis = 1)

preform,t = main_nn(6000,5,chebyshev)
print ('Chebyshev distance: Acc = {:.1f}%, TimeElapsed = {:e}ms'.format(preform * 100, t * 1000))

preform,t = main_nn(6000,5,chebyshev)
print ('Manhattan distance: Acc = {:.1f}%, TimeElapsed = {:e}ms'.format(preform * 100, t * 1000))

preform,t = main_nn(6000,5,minkowski)
print ('Euclid distance: Acc = {:.1f}%, TimeElapsed = {:e}ms'.format(preform * 100, t * 1000))

Chebyshev distance: Acc = 55.7%, TimeElapsed = 8.415920e+04ms
Manhattan distance: Acc = 65.7%, TimeElapsed = 8.348229e+04ms
Euclid distance: Acc = 93.7%, TimeElapsed = 1.313823e+05ms


### Finding the weight for each element

We are about to find $\vec a$, s.t. $x_k' = a_kx_k$ performs better than the original one. This is equivalence to finding a diagonal, positive definite matrix $\mathbf Q$ such that the euclid distance would change to $\vec x^{\mathrm T}\mathbf Q\vec x$

In [27]:
a = np.ones(28*28)
def sqr_distance(test,data, p = 2):
    r_pos = (data-test.reshape((1,-1))) * a
    return np.linalg.norm(r_pos,axis = 1,ord = p)

preform,t = main_nn(6000,5,sqr_distance)
print ('Euclid distance: Acc = {:.1f}%, TimeElapsed = {:e}ms'.format(preform * 100, t * 1000))

Euclid distance: Acc = 93.7%, TimeElapsed = 9.777281e+05ms
